bigtable/client/row_range_test.cc

```diff
@@ -0,0 +1,115 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/row_range.h"
+
+#include <gmock/gmock.h>
+
+TEST(RowRangeTest, Equal) {
+  const std::string keys[] = {
+    "",
+    "a",
+    "b",
+    "c",
+  };
+
+  int size = sizeof(keys)/sizeof(*keys);
+
+  // Test non-empty combinations (x,y where x!=y) of the keys above.
+  for (int i = 0; i < size; i++) {
+    for (int j = i+1; j < size; j++) {
+      for (int k = 0; k < size; k++) {
+        for (int l = k+1; l < size; l++) {
+          bigtable::RowRange r1(keys[i], keys[j]);
+          bigtable::RowRange r2(keys[k], keys[l]);
+
+          if ((keys[i] == keys[k]) && (keys[j] == keys[l])) {
+            // Ranges must be the same
+            ASSERT_TRUE(  (r1 == r2)) << ": " << r1 << "/" << r2;
+            ASSERT_TRUE(! (r1 != r2)) << ": " << r1 << "/" << r2;
+          } else {
+            // Ranges must be diffetent
+            ASSERT_TRUE(! (r1 == r2)) << ": " << r1 << "/" << r2;
+            ASSERT_TRUE(  (r1 != r2)) << ": " << r1 << "/" << r2;
+          }
+        }
+      }
+    }
+  }
+  // Empty range.
+  ASSERT_TRUE(bigtable::RowRange("a", "a") == bigtable::RowRange("b", "b"));
+  // Infinite range ! = Empty range.
+  ASSERT_TRUE(bigtable::RowRange("", "") != bigtable::RowRange("b", "b"));
+  // Infinite ranges.
+  ASSERT_TRUE(bigtable::RowRange("", "") == bigtable::RowRange("", ""));
+}
+
+TEST(RowRangeTest, Prefix) {
+  EXPECT_EQ(bigtable::RowRange::prefix_range("foo"),
+            bigtable::RowRange("foo", "fop"));
+  EXPECT_EQ(bigtable::RowRange::prefix_range(""),
+            bigtable::RowRange("", ""));
+  EXPECT_EQ(bigtable::RowRange::prefix_range("foo\377\377"),
+            bigtable::RowRange("foo\377\377", "fop"));
+  EXPECT_EQ(bigtable::RowRange::prefix_range("\377\377"),
+            bigtable::RowRange("\377\377", ""));
+}
+
+TEST(RowRangeTest, RowRanges) {
+  EXPECT_EQ(bigtable::RowRange("fop", "fox"),
+            bigtable::RowRange::open_interval("foo", "fox"));
+  EXPECT_EQ(bigtable::RowRange("foo", "foy"),
+            bigtable::RowRange::closed_interval("foo", "fox"));
+  EXPECT_EQ(bigtable::RowRange("fop", "foy"),
+            bigtable::RowRange::left_open("foo", "fox"));
+  EXPECT_EQ(bigtable::RowRange("foo", "fox"),
+            bigtable::RowRange::right_open("foo", "fox"));
+  EXPECT_EQ(bigtable::RowRange("foo", "fox"),
+            bigtable::RowRange::range("foo", "fox"));
+  EXPECT_EQ(bigtable::RowRange("foo", "foo"),
+            bigtable::RowRange::empty());
+}
+
+static std::string Intersect(const std::string& s1, const std::string& l1,
```

Comment:
The rest of the code in the project is using `std::string const&`, or more generally `T const&`.